include/boost/gil/rasterization/line.hpp

```diff
@@ -0,0 +1,65 @@
+#include <boost/gil/point.hpp>
+#include <cmath>
+#include <cstddef>
+
+namespace boost { namespace gil {
+/// \defgroup Rasterization
+/// \brief A set of functions to rasterize shapes
+///
+/// Due to images being discrete, most shapes require specialized algorithms to handle rasterization
+/// efficiently and solve problem of connectivity and being close to the original shape.
+
+/// \defgroup LineRasterization
+/// \ingroup Rasterization
+/// \brief A set of rasterizers for lines
+///
+/// The main problem with line rasterization is to do it efficiently, e.g. less floating point
+/// operations. There are multiple algorithms that on paper should reach the same result, but
+/// due to quirks of IEEE-754 they don't. Please select one and stick to it if possible. At
+/// the moment only Bresenham rasterizer is implemented.
+
+/// \ingroup LineRasterization
+/// \brief Rasterize a line according to Bresenham algorithm
+///
+/// Do note that if either width or height is 1, slope is set to zero.
+/// reference:
+/// https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm#:~:text=Bresenham's%20line%20algorithm%20is%20a,straight%20line%20between%20two%20points.
+struct bresenham_line_rasterizer
+{
+    constexpr std::ptrdiff_t point_count(std::ptrdiff_t width, std::ptrdiff_t height) const noexcept
+    {
+        return width > height ? width : height;
+    }
+
+    template <typename RAIterator>
+    void operator()(std::ptrdiff_t width, std::ptrdiff_t height, RAIterator output_first) const
```

Comment:
1. Why the rasterizer does not take coordinates of from point and to point?
    I'm not clear why it has the fixed `point_t(0, 0);` later below.

2. Why this is a function object? Why not `line_rasterizer` taking `bresenham_rasterizer` strategy type and applied from start point to end point? Public function overloads

```cpp
rasterize_line(from, to, output);
output = rasterize_line(from, to);
```

Reply:
1. That interface is probably better, yes. I don't really remember why I chose to go with `width` and `height`. I will rework this tomorrow.

2. I wanted it to be easy to pass into function templates. Circle rasterizers for example are passed into Hough transform for circles.

Reply:
> 2. I wanted it to be easy to pass into function templates.
>     Circle rasterizers for example are passed into Hough transform for circles.

If rasteriser is dedicated as public functionality for users, I'd consider its direct call as primary use case and design its interfaces with end-user convenience in mind.
Then, passing the rasterisers into other functions would be a secondary use-case - it can be wrapped with function-local `struct` or lambda.

However, we are not required to release the rasterisers as part of the public interfaces now - our focus is on the Hough transforms here. You can just move it into `boost::gil::detail` and keep it as it is. W can release the rasterisers after we polish their interface, in future.

Perhaps @lpranam has an advice/opinion on that.

Reply:
It was much harder than I thought, but I implemented two point version of the interface. Hough transform for lines tests just worked after small modifications.

Reply:
The testing is has breadth and depth, but there is a design decision problem: current implementation does not guarantee equivalence of a line drawn in different direction to the current one. I could use canonical direction, but am not sure if it is worth it. See the tests for more context. I believe now it is usable, but I guess some more views wouldn't hurt.

Reply:
As for the filling the pixels themselves not being one step process with rasterization, we could add something like this:

    template <typename View, typename Rasterizer, typename ... RasterizerArgs>
    void render(View const& view, typename View::value_type fill_pixel, Rasterizer& rasterizer, RasterizerArgs&& ... args) {
        // some fancy iterator that fills on *it = point
        auto iterator = make_fill_iterator(view);
        rasterizer(iterator, args...);
    }

Overloads for circle rasterizers would be useful so could sort the output by scanlines first, otherwise they otherwise destroy cache comprehension.